File: 2023/archive/src/day1.kt
import java.io.File

val NUMBER_TO_DIGIT_MAP =
    mapOf(
        "one" to 1,
        "two" to 2,
        "three" to 3,
        "four" to 4,
        "five" to 5,
        "six" to 6,
        "seven" to 7,
        "eight" to 8,
        "nine" to 9,
        "1" to 1,
        "2" to 2,
        "3" to 3,
        "4" to 4,
        "5" to 5,
        "6" to 6,
        "7" to 7,
        "8" to 8,
        "9" to 9
    )

var part1RunningTotal = 0
var part2RunningTotal = 0

fun computePart1(line: String): Int {
  val firstDigit = line.filter { it.isDigit() }.firstOrNull()
  val lastDigit = line.reversed().filter { it.isDigit() }.firstOrNull()

  return "$firstDigit$lastDigit".toInt()
}

fun computePart2(line: String): Int {
  val firstMatch = line.findAnyOf(NUMBER_TO_DIGIT_MAP.keys)
  val lastMatch = line.findLastAnyOf(NUMBER_TO_DIGIT_MAP.keys)

  val firstDigit = NUMBER_TO_DIGIT_MAP[firstMatch!!.second]
  val lastDigit = NUMBER_TO_DIGIT_MAP[lastMatch!!.second]

  return "$firstDigit$lastDigit".toInt()
}

fun main() {
  val lineList = mutableListOf<String>()

  // The File() call is relative to the directory we're running run.sh from
  File("./input/day1.txt").useLines { lines -> lines.forEach { lineList.add(it) } }

  lineList.forEach { line ->
    part1RunningTotal += computePart1(line)
    part2RunningTotal += computePart2(line)
  }

  println("Part 1 total: $part1RunningTotal")
  println("Part 2 total: $part2RunningTotal")
}

File: 2023/archive/src/day1.rb
# day1.rb - Day 1 solution, ruby-fied!
class Day1
  @@number_to_digit_hash = {
    'one' => 1,
    'two' => 2,
    'three' => 3,
    'four' => 4,
    'five' => 5,
    'six' => 6,
    'seven' => 7,
    'eight' => 8,
    'nine' => 9,
    '1' => 1,
    '2' => 2,
    '3' => 3,
    '4' => 4,
    '5' => 5,
    '6' => 6,
    '7' => 7,
    '8' => 8,
    '9' => 9
  }

  def initialize(input_lines)
    @input_lines = input_lines
  end

  def part_1
    @input_lines.sum do |line|
      if line[0] == "#"
        next
      end

      all_digits = line.scan(/\d/)
      "#{all_digits.first}#{all_digits.last}".to_i
    end
  end

  def part_2
    @input_lines.sum do |line|
      if line[0] == "#"
        next
      end

      all_digits = line.scan(/(?=(\d|one|two|three|four|five|six|seven|eight|nine))/).flatten
      "#{@@number_to_digit_hash[all_digits.first]}#{@@number_to_digit_hash[all_digits.last]}".to_i
    end
  end
end

day_1 = Day1.new(IO.readlines("./input/day1.txt"))
puts "Part 1 total: #{day_1.part_1}" # Part 1 total: 54331
puts "Part 2 total: #{day_1.part_2}" # Part 2 total: 54518
